damus/Util/Zap.swift

```diff
@@ -0,0 +1,307 @@
+        if tag.count >= 2 && tag[0] == name {
+            return tag[1]
+        }
+    }
+    
+    return nil
+}
+
+func decode_zap_request(_ desc: String) -> ZapRequest? {
+    let decoder = JSONDecoder()
+    guard let jsonData = desc.data(using: .utf8) else {
+        return nil
+    }
+    guard let jsonArray = try? JSONSerialization.jsonObject(with: jsonData) as? [[Any]] else {
+        return nil
+    }
+    
+    for array in jsonArray {
+        guard array.count == 2 else {
+            continue
+        }
+        let mkey = array.first.flatMap { $0 as? String }
+        if let key = mkey, key == "application/nostr" {
+            guard let dat = try? JSONSerialization.data(withJSONObject: array[1], options: []) else {
+                return nil
+            }
+            
+            guard let zap_req = try? decoder.decode(NostrEvent.self, from: dat) else {
+                return nil
+            }
+            
+            guard zap_req.kind == 9734 else {
+                return nil
+            }
+            
+            /// Ensure the signature on the zap request is correct
+            guard case .ok = validate_event(ev: zap_req) else {
+                return nil
+            }
+            
+            return ZapRequest(ev: zap_req)
+        }
+    }
+    
+    return nil
+}
+
+
+
+func fetch_zapper_from_lnurl(_ lnurl: String) async -> String? {
+    guard let endpoint = await fetch_static_payreq(lnurl) else {
+        return nil
+    }
+    
+    guard let allows = endpoint.allowsNostr, allows else {
+        return nil
+    }
+    
+    guard let key = endpoint.nostrPubkey, key.count == 64 else {
+        return nil
+    }
+    
+    return endpoint.nostrPubkey
+}
+
+func decode_lnurl(_ lnurl: String) -> URL? {
+    guard let decoded = try? bech32_decode(lnurl) else {
+        return nil
+    }
+    guard decoded.hrp == "lnurl" else {
+        return nil
+    }
+    guard let url = URL(string: String(decoding: decoded.data, as: UTF8.self)) else {
+        return nil
+    }
+    return url
+}
+
+func fetch_static_payreq(_ lnurl: String) async -> LNUrlPayRequest? {
+    guard let url = decode_lnurl(lnurl) else {
+        return nil
+    }
+    
+    guard let ret = try? await URLSession.shared.data(from: url) else {
+        return nil
+    }
+    
+    let json_str = String(decoding: ret.0, as: UTF8.self)
+    
+    guard let endpoint: LNUrlPayRequest = decode_json(json_str) else {
+        return nil
+    }
+    
+    return endpoint
+}
+
+func fetch_zap_invoice(_ payreq: LNUrlPayRequest, zapreq: NostrEvent, amount: Int64) async -> String? {
+    guard var base_url = payreq.callback.flatMap({ URLComponents(string: $0) }) else {
+        return nil
+    }
+    
+    let zappable = payreq.allowsNostr ?? false
+    
+    var query = [URLQueryItem(name: "amount", value: "\(amount)")]
+    
+    if zappable {
+        if let json = encode_json(zapreq) {
+            query.append(URLQueryItem(name: "nostr", value: json))
```

Comment:
This will result in a different flow than what is expected for LNURL, afaik the description hash must match what the `payService` responds with, in which case we would have to return the tip note there

Reply:
I think its fine, wallets will not be implementing this. I think it only makes sense with nostr clients.